src/foam/dao/index/AutoIndex.js

```diff
@@ -20,57 +20,191 @@
 foam.CLASS({
   package: 'foam.dao.index',
   name: 'AutoIndex',
-  extends: 'foam.dao.index.Index',
+  extends: 'foam.dao.index.ProxyIndex',
 
   requires: [
     'foam.core.Property',
-    'foam.dao.index.NoPlan'
+    'foam.dao.index.NoPlan',
+    'foam.dao.index.CustomPlan',
+    'foam.mlang.predicate.And',
+    'foam.mlang.predicate.Or',
+    'foam.dao.index.AltIndex',
+    'foam.dao.index.ValueIndex',
+    'foam.mlang.predicate.True',
+    'foam.mlang.predicate.False',
   ],
 
+  constants: {
+    /** Maximum cost for a plan which is good enough to not bother looking at the rest. */
+    GOOD_ENOUGH_PLAN: 20
+  },
+
   properties: [
     {
-      name: 'properties',
-      factory: function() { return {}; }
+      /** Used to create the delegate ID index for new instances of AutoIndex */
+      name: 'idIndex',
+      required: true
     },
     {
-      name: 'mdao'
+      name: 'delegate',
+      factory: function() {
+        return this.AltIndex.create({ delegates: [ this.idIndex ] });
+      }
     }
   ],
 
   methods: [
-    function put() { },
+    function estimate(size, sink, skip, limit, order, predicate) {
+      return this.delegate.estimate(size, sink, skip, limit, order, predicate);
+    },
+
+    function toPrettyString(indent) {
+      var ret = "";
```

Comment:
Google and FOAM standard is to use single quotes for string literals unless string contains a single quote.